### DISCUSSION TITLE
Input Subsystem

### DISCUSSION THREAD

**Discussion by gnab-gib-se** [endless-sky/endless-sky#8613] [Balance, Mechanics, and Interactions]
I'm a professional C++ software engineer and I'm interested in contributing to this project.  I've been reviewing the backlog and noticed quite a few issues related to how input is handled that caught my attention:

Open
#13 Better support for French AZERTY (and other keyboards)
#2291 Switchable Hotkey Profiles
#3972 Controller integration
#5114 Allow binding of shortcuts and/or support alternative keyboard layouts
#5954 After keyboard layout switch keybindings stop working
#7132 feat(UI) Gamepad controller support
#8331 [Bug] Hidden keybinding for zoom
#8334 feat(ui): Unbind Keybinds

Closed/Merged
#1527 Preference toggle to always show the button underlines
#2208 interface hotkeys configuration
#2393 Cannot make groups for tactical and strategic organization of fleets
#3846 Joystick Support
#4944 zoom-in not working.
#5624 make zoom work with numpad keys
#5724 Request: Change default keybinds to WASD-based
#6936 Added option for firing and rotating the ship with mouse

I started investigating the code and considering why some of these have been outstanding so long (especially #13) and what would be required to enable fixes and new features here.

There is no input subsystem and it is being handled in a number of different ways, from different places, and even from different threads.  There are several comments throughout UI and Engine about why this or that input has to be checked here or there to "avoid race conditions" rather than a cohesive strategy to prevent them (e.g. handle queued inputs(events)->calculate->redraw).  There are a lot of assumptions (ASCII), hard-coded keycodes, and mixing of keycodes and scancodes which will certainly cause problems for other keyboard layouts.

As a result of all of the above I suspect the area of input handling is fragile.  Care must be taken to avoid breaking the delicate balance that exists today between UI and Engine while moving toward a unified input model.

#7132 is unfortunate as it perfectly demonstrates (and describes) the same issues I'm seeing, but rather than starting by refactoring to correct the underlying impediments, it just layered even more complexity on top.

A logical first step would be some refactoring to consolidate input code and separate it from game logic.

I will stop there as I think this is enough to start a discussion.  Thoughts?  Agree?  Disagree?  What kind of appetite would there be for contributions in this area?

**Comment by quyykk** [comment]
>  What kind of appetite would there be for contributions in this area?

Yes revamping the input handling system of the game would be a very good idea. I'm not sure how feasible it is to do incremental improvements, but if it is possible I would highly recommend doing that, for easier reviewing (since we're kinda swamped). TBH I haven't given much thought to how such an better input handling system might look like, but this issue would be the perfect place to discuss it before doing any work :smile:.

**Comment by gnab-gib-se** [comment]
Yeah, I definitely agree with incremental improvements; it's difficult to imagine how else you could get anything done in this code base with so many contributions.

I've been thinking about this quite a bit (as you'll see), and here are some ideas.

<details>
<summary>

## Philosophy

</summary>
<p>
First, we should be clear on why we want to do this. After all, SDL already provides multiple input abstractions. The core issue is that the input abstractions SDL provides aren't very rich. Since it doesn't handle things like bindings, keyboard layouts, or easily switching between different input control schemes, it's desirable to add another layer of abstraction that can support all that. This can all be done with SDL (and indeed will be, behind the scenes), but it will make the game code much easier to read and more maintainable to have an input subsystem that puts things into terms the game code understands. Hopefully this addresses any question like "Why don't we just fix the usage of SDL?".
</p>
<p>
The primary goal of such a subsystem would be to abstract many of the input details that are currently spread liberally throughout the game code (SDLK_*, SDL_Keycode) and instead keep the game code focused on what it actually cares about. That is, something wants the ship to stop, not whether someone pressed Shift+BACK or not. The mapping between an input like Shift+Down and notifying the game code that a stop was requested is the business of an input subsystem. The game code only needs to know when a particular game function is called for, based on some input. The game code doesn't need to know what input it was or any details about how we received that input.
</p>
<p>
Another key point is that we don't need to reinvent the wheel. There are already many input subsystems we can look to for inspiration in other engines that provide richer abstractions. Of course, this is not an engine, so only a small subset of the features are needed, but the way that things are organized can be very instructive because many of the goals are the same. A side benefit of this is that it will make it much easier to port to another, more modern engine in the future if that is ever desired.
</p>
</details>

<details>
<summary>

## Starting Point

</summary>
<p>
The existing keybindings are tightly coupled to the Command class, which is the source of some of the issues I've mentioned as well as other problematic assumptions. Extending Command is not a good idea, and instead I favor an approach that separates the concerns of input and keybindings from it.
</p>
<p>
In earlier iterations, the input subsystem could generate Command objects where necessary, inserting itself between where inputs are read today and being interpreted into Commands. In later iterations, the Command class could probably be replaced, eliminated entirely, or significantly refactored from its current state.
</p>
<p>
At a high level, I would like to add a layer of input processing between each place where we get input from facilities like SDL_Event (UI::Handle) and SDL_GetKeyboardState (Engine::HandleKeyboardInputs -> Command::ReadKeyboard) and the game code that results.
</p>
<p>
A good starting point for conceptualizing what this could look like is the low-hanging fruit of the shortcuts throughout the UI panels. These are currently not bindable, relatively independent, and have a low impact and low risk of refactoring. I will focus on these for this example, but I will also include some notional information on more advanced axis controls just to show how and where those could later be integrated into this model.
</p>
</details>

<details>
<summary>

## High-Level Design

</summary>
<p>
The visitor pattern is a very good fit for this problem, allowing us to have a dual inheritance hierarchy (to separate input from mappings and bindings). However, it can still be bound together and maps very well to the overall "if else if" pattern of much of the existing input handling. Each input can visit each mapping and its associated bindings, starting at the top panel and working its way down until it is handled by a matching binding or none is found.
</p>
<p>
This involves turning SDL_Event (and potentially other inputs) into an InputEvent object, which could then be matched against various possible InputMapping objects (which contain InputBinding objects) on a given Panel that are built from configuration. If any of the bindings accept the input, they execute the InputMapping (the closest analogy to a Command in this new model), which all have a callback function that can run a block of code directly or generate and queue a Command.
</p>
<p>
Here is some pseudocode for a detailed example of how this could look for the MainPanel Zoom mappings, with the following caveats:
</p>

- no const-correctness
- no pointer or reference syntax (unless it helps to communicate specific ideas)
- ... but const-reference will be preferred where ever possible
- no scoping or accessors
- ... but data members would all be private with only the necessary accessors
- no semicolons
- style intentionally compact; I know the braces don't match the style used in the code
</details>

<p>
This is probably more detailed than it needed to be, but I added quite a few comments to aid understanding. This is a complete example showing how buttons and axis controls can be elegantly combined without the game code knowing about it. I also recognize that what I'm proposing is a fairly substantial refactor, so I wanted to show up front that I've put a significant amount of thought into this before jumping into coding something.
</p>

### Panels
```
class Panel : InputEventHandler
	// This would eventually replace KeyDown entirely.
	//
	// Mappings that are truly common to every Panel could potentially be
	// implemented here, e.g., an ESC menu, a console, etc.
	virtual bool HandleInputEvent(const InputEvent &) overrides {
		return false
	}

class MainPanel : Panel
	// The MainPanel Zoom is a particularly interesting example because it's
	// currently implemented as both buttons/actions (MainPanel::KeyDown) and
	// an axis (MainPanel::Scroll).
	//
	// One of the major advantages of an input subsystem is the ability to
	// represent both types of bindings as a single 1D axis in the game code.
	//
	// This would also be an ideal way to represent the movement keys so that
	// they can easily be mapped to various button/action combinations (WASD,
	// arrows) or axis (mouse, controller, joystick) inputs without further
	// complicating the game code.

	// PLACEHOLDER EXAMPLE: Zoom as refactored buttons/actions only (just to
	// show what the existing buttons would look like implemented this way),
	// but this mechanism would be more appropriate for, e.g., the 0-9
	// SelectGroup keys, which are true buttons/actions and not axes.
	InputActionMapping zoomViewOutActionMapping = InputActionMapping(typeid(*this).name(), "Zoom view out",
		InputCallback(std::bind(&MainPanel::ZoomViewOut, this)))
	InputActionMapping zoomViewInActionMapping = InputActionMapping(typeid(*this).name(), "Zoom view in",
		InputCallback(std::bind(&MainPanel::ZoomViewIn, this)))
	InputContext inputContext(zoomViewOutActionMapping, zoomViewInActionMapping)

	bool ZoomViewOut() {
		Preferences::ZoomViewOut()
		return true
	}

	bool ZoomViewIn() {
		Preferences::ZoomViewIn()
		return true
	}

	// OR

	// This is a better way to implement Zoom directly as a 1D axis, which can
	// have both button/action and axis bindings underneath.
	InputAxis1dMapping zoomViewAxis1dMapping = InputAxis1dMapping(typeid(*this).name(), "Zoom view",
		InputCallback(std::bind(&MainPanel::ZoomView, this, _1)))
	InputContext inputContext(zoomViewAxis1dMapping)

	bool ZoomView(float delta) {
		if(delta < 0)
			Preferences::ZoomViewOut()
		else if(delta > 0)
			Preferences::ZoomViewIn()
		else
			return false

		return true
	}


	virtual bool HandleInputEvent(const InputEvent &inputEvent) override {
		return inputContext.HandleInputEvent(inputEvent)
	}
```

### Usage
```
bool UI::Handle(const SDL_Event &event)
{
	shared_ptr<InputEvent> inputEvent = InputEvent::Create(event)
		...
		else if(inputEvent)
		{
			handled = (*it)->HandleInputEvent(*inputEvent.get())
		}
		...
}
```

As you can see above, all of the details about keycodes, layouts, and indeed whether this is even a keyboard binding at all are hidden from MainPanel and the UI. This scheme could be extended to include 2D axis input as well (e.g., (x,y) from a mouse). It would be straightforward to support a MouseButtonInput/MouseButtonBinding to enable the use of extra mouse buttons for simple actions (like Zooming with SDL_BUTTON_X1 and SDL_BUTTON_X2).

You can then radically change the internals of KeyInput/KeyBinding and MouseWheelInput/MouseWheel1dBinding (in the input system implementation below) to Solve All The Problems©®™ without changing any game code.

<details>
<summary>

### Hypothetical Simulated Events Example

</summary>

```
// PLACEHOLDER EXAMPLE: In practice, these events would be generated from,
// e.g., UI::Handle with no other references to SDL as above. However, for the
// sake of demonstration, it might help to make sense of it with a
// hypothetical SDL_Event constructor like this:

// zoom out
MainPanel::HandleInputEvent(InputEvent::Create(SDL_Event(SDL_KEYDOWN, SDLK_MINUS)))
// or
MainPanel::HandleInputEvent(InputEvent::Create(SDL_Event(SDL_KEYDOWN, SDLK_KP_MINUS)))

// zoom in
MainPanel::HandleInputEvent(InputEvent::Create(SDL_Event(SDL_KEYDOWN, SDLK_PLUS)))
// or
MainPanel::HandleInputEvent(InputEvent::Create(SDL_Event(SDL_KEYDOWN, SDLK_KP_PLUS)))
// or
MainPanel::HandleInputEvent(InputEvent::Create(SDL_Event(SDL_KEYDOWN, SDLK_EQUALS)))
```
</details>

Everything that follows is just a more detailed breakdown of all the input subsystem code that would be needed to make the above possible.

<details>
<summary>

### Events

</summary>

<p>
Events are wrappers around SDL_Event or any other kind of input (such as simulated or test input).
</p>

```
class InputEventHandler
	// Check the specified InputEvent looking for a match and handle the event
	// if one is found.
	// Return true if a match was found and the event was handled.
	virtual bool HandleInputEvent(const InputEvent &) = 0

// A single discrete input, 1:1 with an SDL_Event.
class InputEvent
	SDL_Event event

	static shared_ptr<InputEvent> Create(SDL_Event event) {
		if(event.type == SDL_KEYDOWN || event.type == SDL_KEYUP)
			return new KeyInput(event)
		else if(event.type == SDL_MOUSEWHEEL)
			return new MouseWheelInput(event)
		else
			return nullptr
	}

	// These default implementations are the first layer for deciding whether
	// a given input can match a given binding without needing to downcast.

	// ================================
	// InputActionEvent Visit functions
	// ================================

	// SDL_KeyboardEvent (SDL_KEYDOWN, SDL_KEYUP)
	virtual bool Visit(KeyBinding) {
		return false
	}

	// PLACEHOLDER EXAMPLE: The following would not be in the initial
	// iterations, just showing some examples of how it could be extended.

	// SDL_ControllerButtonEvent (SDL_CONTROLLERBUTTONDOWN, SDL_CONTROLLERBUTTONUP)
	virtual bool Visit(ControllerButtonBinding) {
		return false
	}

	// SDL_JoyButtonEvent (SDL_JOYBUTTONDOWN, SDL_JOYBUTTONUP)
	virtual bool Visit(JoyButtonBinding) {
		return false
	}

	// ==============================
	// InputAxisEvent Visit functions
	// ==============================

	// SDL_MouseWheelEvent (SDL_MOUSEWHEEL)
	virtual bool Visit(MouseWheel1dBinding) {
		return false
	}

	// PLACEHOLDER EXAMPLE: The following would not be in the initial
	// iterations, just showing some examples of how it could be extended.

	// SDL_MouseMotionEvent (SDL_MOUSEMOTION)
	virtual bool Visit(MouseMotionBinding) {
		return false
	}

	// SDL_MouseButtonEvent (SDL_MOUSEBUTTONDOWN, SDL_MOUSEBUTTONUP)
	virtual bool Visit(MouseButtonBinding) {
		return false
	}

// The subset of SDL_Event that deals with boolean input, typically a key or
// button, which can be down/pressed or up/released. These are also sometimes
// referred to as digital, binary, action, or trigger inputs. We combine the
// SDL DOWN/UP events and will let the InputBinding determine which they
// execute on.
class InputActionEvent : InputEvent

class KeyInput : InputActionEvent
	KeyInput(SDL_Event event) {
		this->event = event
	}

	virtual bool Visit(KeyBinding keyBinding) override {
		// This inversion of control is not strictly necessary, but InputEvent
		// objects will probably be simpler, so it will simplify the
		// interaction between these objects to allow the InputBinding objects
		// to do the heavy lifting.
		return keyBinding.MatchAndExecute(this)
	}

	SDL_Keycode GetKeycode() {
		return this->event.key.keysym.sym
	}

	bool IsDown() {
		return this->event.type == SDL_KEYDOWN
	}

	bool IsUp() {
		return this->event.type == SDL_KEYUP
	}

// The subset of SDL_Event that deals with axis input, typically a mouse,
// controller, joystick, or touch, e.g.
// 1D
// SDL_MouseWheelEvent (SDL_MOUSEWHEEL) (technically 2D, but frequently used 1D)
// SDL_JoyAxisEvent (SDL_JOYAXISMOTION)
// SDL_ControllerAxisEvent (SDL_CONTROLLERAXISMOTION)
//
// 2D
// SDL_MouseWheelEvent (SDL_MOUSEWHEEL)
// SDL_MouseMotionEvent (SDL_MOUSEMOTION)
// SDL_MouseButtonEvent (SDL_MOUSEBUTTONDOWN, SDL_MOUSEBUTTONUP)
// SDL_JoyBallEvent (SDL_JOYBALLMOTION)
// SDL_TouchFingerEvent (SDL_FINGERMOTION, SDL_FINGERDOWN, SDL_FINGERUP)
class InputAxisEvent : InputEvent

class MouseWheelInput : InputAxisEvent
	MouseWheelInput(SDL_Event event) {
		this->event = event
	}

	virtual bool Visit(MouseWheel1dBinding mouseWheel1dBinding) override {
		return mouseWheel1dBinding.MatchAndExecute(this)
	}

	float GetPreciseX() {
		return this->event.wheel.preciseX
	}

	float GetPreciseY() {
		return this->event.wheel.preciseY
	}
```
</details>

<details>
<summary>

### Bindings

</summary>

<p>
Bindings are what we typically think of as keybinds plus associating axes on devices like mice, controllers, and joysticks.
</p>

```
class InputBinding
	InputCallback callback
	float scale

	// Return true if this binding will accept the given input and a mapping
	// should be executed.
	virtual bool Accept(const InputEvent &) = 0

class InputActionBinding : InputBinding
	TriggerType triggerType

	// This is not a bool to allow for future extensibility (double-click,
	// long press, triggering on both, repeat, etc.)
	enum TriggerType
		TRIGGER_TYPE_DOWN
		TRIGGER_TYPE_UP

class KeyBinding : InputActionBinding
	SDL_Keycode keycode

	KeyBinding(InputCallback callback, InputActionBinding::TriggerType triggerType, SDL_Keycode keycode, float scale = 1.0) {
		this->callback = callback
		this->triggerType = triggerType
		this->keycode = keycode
		this->scale = scale
	}

	virtual bool Accept(const InputEvent &inputEvent) override {
		return inputEvent.Visit(this)
	}

	// There is probably an opportunity to implement this generically at the
	// InputActionEvent/InputActionBinding level with some abstractions.
	bool MatchAndExecute(KeyInput keyInput) {
		if(keyInput.GetKeycode() == this->keycode)
			if(keyInput.IsDown() && this->triggerType == InputActionBinding::TRIGGER_TYPE_DOWN) ||
				(keyInput.IsUp() && this->triggerType == InputActionBinding::TRIGGER_TYPE_UP)
				return this->callback.Execute(this->scale)
		return false
	}

class InputAxisBinding : InputBinding

class InputAxis1dBinding : InputAxisBinding
	Axis axis

	// Used to select an axis when using a 2D input as 1D.
	enum Axis
		AXIS_X
		AXIS_Y

class MouseWheel1dBinding : InputAxis1dBinding
	InputAxis1dBinding::Axis axis

	MouseWheel1dBinding(InputCallback callback, InputAxis1dBinding::Axis axis, float scale) {
		this->callback = callback
		this->axis = axis
		this->scale = scale
	}

	virtual bool Accept(const InputEvent &inputEvent) override {
		return inputEvent.Visit(this)
	}

	bool MatchAndExecute(MouseWheelInput mouseWheelInput) {
		if(this->axis == AXIS_X)
			if(mouseWheelInput.GetPreciseX() > 0)
				return this->callback.Execute(this->scale * mouseWheelInput.GetPreciseX())
		else if(this->axis == AXIS_Y)
			if(mouseWheelInput.GetPreciseY() > 0)
				return this->callback.Execute(this->scale * mouseWheelInput.GetPreciseY())
		else
			assert false
		
		return false
	}

class InputAxis2dBinding : InputAxisBinding

class MouseMotionBinding : InputAxis2dBinding
...
```
</details>

<details>
<summary>

### Mappings

</summary>

<p>
Mappings are game functions that can be mapped to zero or more bindings. I'm not thrilled with this name, but I went with it for now as this is the terminology that the Unreal Engine uses for the same concept, and I mostly followed those conventions.
</p>

```
// This class provides a simple abstraction for the different types of
// callback functions that mappings can have.
class InputCallback
	std::function<bool()> action_callback
	std::function<bool(float)> axis1d_callback
	std::function<bool(float, float)> axis2d_callback

	InputCallback(std::function<bool()> action_callback)
		this->action_callback = action_callback
		this->axis1d_callback = nullptr
		this->axis2d_callback = nullptr

	InputCallback(std::function<bool(float)> axis1d_callback)
		this->action_callback = nullptr
		this->axis1d_callback = axis1d_callback
		this->axis2d_callback = nullptr

	InputCallback(std::function<bool(float, float)> axis2d_callback)
		this->action_callback = nullptr
		this->axis1d_callback = nullptr
		this->axis2d_callback = axis2d_callback

	// There will always be at least one argument provided (scale), though it
	// may not be used. For this reason there is no need for a no argument
	// Execute method.

	// It's OK to pass fewer arguments to the callback if the mapping doesn't
	// require them, but never more; this allows, e.g., an axis1d or axis2d
	// binding to be used with an action mapping that requires no parameters
	// (e.g., clicking a mouse button, but not caring where the click
	// occurred.)
	bool Execute(float delta) {
		if(action_callback)
			return action_callback()
		else if(axis1d_callback)
			return axis1d_callback(delta)
		assert false
	}

	// We don't refer to these as dx and dy because the axes may have been
	// mapped differently.
	bool Execute(float delta1, float delta2) {
		if(action_callback)
			return action_callback()
		else if(axis1d_callback)
			return axis1d_callback(delta1)
		else if(axis2d_callback)
			return axis2d_callback(delta1, delta2)
		assert false
	}

// An InputMapping represents a game function on a particular Panel. It should
// have a collection of InputBinding objects that represent the ways the
// mapping can be executed, so one mapping can have multiple bindings.
class InputMapping : InputEventHandler
	vector<InputBinding> bindings
	// Return true unless you want matching this input against mappings /
	// bindings to continue.
	InputCallback callback

class InputActionMapping : InputMapping
	// panelName is only used to uniquely identify mappingName when multiple panels
	// have similar or the same functions (e.g., M to bring up a map).
	InputActionMapping(string panelName, string mappingName, InputCallback callback) {
		this->callback = callback

		// PLACEHOLDER EXAMPLE: This is where appropriate InputActionBinding
		// objects would be created from config, e.g., inputs.txt.
		//
		// Look up panelName and mappingName in config, and create the
		// appropriate bindings for this mapping.
		// 
		// For this example, just hard-code them with hypothetical inputs.txt
		// entries as comments (you probably want a more structured format):
		// panelName mappingName bindingName triggerType scancode
		if(panelName == "MainPanel" && mappingName == "Zoom view out")
			// "MainPanel" "Zoom view out" KeyBinding TRIGGER_TYPE_DOWN <SDLK_MINUS>
			bindings.push_back(KeyBinding(this->callback, InputActionBinding::TRIGGER_TYPE_DOWN, SDLK_MINUS))
			// "MainPanel" "Zoom view out" KeyBinding TRIGGER_TYPE_DOWN <SDLK_KP_MINUS>
			bindings.push_back(KeyBinding(this->callback, InputActionBinding::TRIGGER_TYPE_DOWN, SDLK_KP_MINUS))
		else if(panelName == "MainPanel" && mappingName == "Zoom view in")
			// "MainPanel" "Zoom view in" KeyBinding TRIGGER_TYPE_DOWN <SDLK_PLUS>
			bindings.push_back(KeyBinding(this->callback, InputActionBinding::TRIGGER_TYPE_DOWN, SDLK_PLUS))
			// "MainPanel" "Zoom view in" KeyBinding TRIGGER_TYPE_DOWN <SDLK_KP_PLUS>
			bindings.push_back(KeyBinding(this->callback, InputActionBinding::TRIGGER_TYPE_DOWN, SDLK_KP_PLUS))
			// "MainPanel" "Zoom view in" KeyBinding TRIGGER_TYPE_DOWN <SDLK_EQUALS>
			bindings.push_back(KeyBinding(this->callback, InputActionBinding::TRIGGER_TYPE_DOWN, SDLK_EQUALS))
	}

	virtual bool HandleInputEvent(const InputEvent &inputEvent) override {
		for(bindings)
			if(it->Accept(inputEvent))
				return true
		return false
	}

class InputAxisMapping : InputMapping

class InputAxis1dMapping : InputAxisMapping
	InputAxis1dMapping(string panelName, string mappingName, InputCallback callback) {
		this->callback = callback

		if(panelName == "MainPanel" && mappingName == "Zoom view")
			// action bindings
			// panelName mappingName bindingName triggerType scancode scale
			// "MainPanel" "Zoom view" KeyBinding TRIGGER_TYPE_DOWN <SDLK_MINUS> -1.0
			bindings.push_back(KeyBinding(this->callback, InputActionBinding::TRIGGER_TYPE_DOWN, SDLK_MINUS, -1.0))
			// "MainPanel" "Zoom view" KeyBinding TRIGGER_TYPE_DOWN <SDLK_KP_MINUS> -1.0
			bindings.push_back(KeyBinding(this->callback, InputActionBinding::TRIGGER_TYPE_DOWN, SDLK_KP_MINUS, -1.0))
			// "MainPanel" "Zoom view" KeyBinding TRIGGER_TYPE_DOWN <SDLK_PLUS> 1.0
			bindings.push_back(KeyBinding(this->callback, InputActionBinding::TRIGGER_TYPE_DOWN, SDLK_PLUS, 1.0))
			// "MainPanel" "Zoom view" KeyBinding TRIGGER_TYPE_DOWN <SDLK_KP_PLUS> 1.0
			bindings.push_back(KeyBinding(this->callback, InputActionBinding::TRIGGER_TYPE_DOWN, SDLK_KP_PLUS, 1.0))
			// "MainPanel" "Zoom view" KeyBinding TRIGGER_TYPE_DOWN <SDLK_EQUALS> 1.0
			bindings.push_back(KeyBinding(this->callback, InputActionBinding::TRIGGER_TYPE_DOWN, SDLK_EQUALS, 1.0))
			// axis bindings
			// panelName mappingName bindingName axis scale
			// "MainPanel" "Zoom view" MouseWheel1dBinding AXIS_Y 1.0
			bindings.push_back(MouseWheel1dBinding(this->callback, InputAxis1dBinding::AXIS_Y, 1.0))
	}

	virtual bool HandleInputEvent(const InputEvent &inputEvent) override {
		for(bindings)
			if(it->Accept(inputEvent))
				return true
		return false
	}

// An InputContext is a grouping of InputMapping objects that should not have
// any overlapping bindings, e.g., all the mappings on a particular Panel.
class InputContext : InputEventHandler
	vector<InputEventHandler> handlers

	InputContext(InputMapping handler...) {
		this->handlers = handler...
	}

	virtual bool HandleInputEvent(const InputEvent &inputEvent) override {
		for(handlers)
			if(it->HandleInputEvent(inputEvent))
				return true
		return false
	}
```
</details>

## Tentative Task List

As far as how to go about doing this incrementally, it could go something like this (which gets less certain the further down you go and can probably be broken down further and parallelized in some cases):

- [ ] PR1: feat(input): Create and test input subsystem infrastructure (that would be most of the classes here under Events, Bindings, and Mappings)
- [ ] PR2: fix(#8331): Modify MainPanel as shown here, exposing this hidden keybinding via inputs.txt (but not yet via Preferences->Controls)
- [ ] PR3-22: refactor(Panel): Each Panel could be refactored and tested as an individual PR (as each of these is completed, it would become possible to change keybindings for that panel via inputs.txt)
[*At this point, #2208 and #5114 could probably be considered completed, available via inputs.txt*]
- [ ] PR23: refactor(PreferencesPanel::DrawControls)!: Make single-column scrollable and searchable, which can support the extra features added to inputs.txt; migrate keys.txt to inputs.txt; and redirect Command::LoadSettings / Command::SaveSettings (ensure #8334 support)
- [ ] PR24: refactor(Command::ReadKeyboard): into MainPanel as Mappings, which generate Command objects and call Command::Set and Command::Clear on Engine::activeCommands (which may need locking or a Command queue, which the Engine can drain at the location it currently calls Command::ReadKeyboard)
[*At this point, all the keyboard inputs in the game should be bindable and handled by the input subsystem. *]
- [ ] PR25: fix(#13)!: keyboard layout support (this has a good chance of affecting how keys are stored in inputs.txt)
- [ ] PR26: fix(#5954): keyboard layout switching
- [ ] PR27: feat(#2291, #5724): Implement switchable input profiles with several example profiles like:
- Default: Current ES profile, Arrows with A afterburner, Tab/Q weapons, M for map
- Escape Velocity: Classic profile, Arrows with Z afterburner, Space/Shift weapons, M for map
- WASD (Derpy): WASD with mostly logical mappings, F afterburner, Space/E weapons, M for map
- WASD (Terin): WASD with mostly positional mappings, Z afterburner, Space/Left Ctrl weapons, Q for map
- [ ] PR28: feat(input): Add complete Axis2d support to the input subsystem
- [ ] PR29: refactor(Panel): axis Mappings using the Axis2d input subsystem (Drag, Hover, ZoneClick, Click, RClick, Release) (should also be splittable by panel)
- [ ] PR30: refactor(Engine): axis Mappings using the Axis2d input subsystem (HandleMouseInput, HandleMouseClicks)
- [ ] PR31: refactor(Engine, UI): Eliminate potential for race conditions by implementing a strict strategy like "handle queued inputs(events)->calculate->redraw" which should be able to unify all the input handling at this point so it's the same everywhere, all gets handled at the same place, and at the same time in the Engine
- [ ] PR32: feat(#7132, #3972): Port Controller support
- [ ] PR33: feat(#3846): Implement Joystick support

EDIT: Corrected some minor mistakes from "refactoring" the pseudocode (it's tough to write a lot of pseudocode without a compiler!)

**Comment by Zireael07** [comment]
Mentioning it here since it's relevant - in that issue where someone wanted to hardcode middle mouse button as a switch for mouse steer, it was mentioned that mouse buttons should be accepted by the keybindings/input system the same way as keypresses

**Reply by gnab-gib-se** [reply]
That can definitely be supported by the above model.

**Reply by gnab-gib-se** [reply]
To elaborate on this a little further, it would work by implementing a MouseButtonInput(based on SDL_MouseButtonEvent (SDL_MOUSEBUTTONDOWN, SDL_MOUSEBUTTONUP)) and a MouseButtonBinding to match it derived from an InputAxis2dBinding.

The MouseButtonInput would contain the SDL_BUTTON_MIDDLE button info as well as x and y coords which would not be used in this case.

This code all exists in Engine right now and is based on alt being held. In the future I would imagine this being handled on MainPanel. Alt down could toggle on, alt up could toggle off, and middle mouse could toggle only on down so it's persistent until you hit it again (which I think is what you're describing).

Hypothetical inputs.txt entries which supports both methods simultaneously:
```
"MainPanel" "Mouse turning toggle" MouseButton2dBinding TRIGGER_TYPE_DOWN BUTTON_TYPE_MIDDLE
"MainPanel" "Mouse turning toggle" KeyBinding TRIGGER_TYPE_DOWN <SDLK_LALT>
"MainPanel" "Mouse turning toggle" KeyBinding TRIGGER_TYPE_DOWN <SDLK_RALT>
"MainPanel" "Mouse turning toggle" KeyBinding TRIGGER_TYPE_UP <SDLK_LALT>
"MainPanel" "Mouse turning toggle" KeyBinding TRIGGER_TYPE_UP <SDLK_RALT>
```

Hypothetical InputActionMapping for this:
```
InputActionMapping mouseTurningToggleActionMapping = InputActionMapping(typeid(*this).name(), "Mouse turning toggle",
		InputCallback(std::bind(&MainPanel::MouseTurningToggle, this)))
```

That InputActionMapping would look up "Mouse turning toggle" in inputs.txt and construct 5 InputBindings internally (1 MouseButton2dBinding, and 4 KeyBinding).  When any of those inputs matched on the MainPanel they would trigger the MainPanel::MouseTurningToggle which would likely generate a Command::MOUSE_TURNING_TOGGLE and queue it for the Engine.

**Comment by Zitchas** [comment]
I can't offer any real insight into the specifics, but some improvement to the way inputs are handled would definitely be appreciated. Thank you for your interest!

**Comment by gnab-gib-se** [comment]
So @quyykk and @Zitchas, you guys both seem to be quite active in the project, who else do you think should give an opinion here before I actually start working on this?  It's a little hard to tell who the key devs are to have on board, but I'd guess some subset of these recently active folks would be good:

warp-core, Amazinite, petervdmeer, Hurleveur, Terin, samrocketman

or some of these less active/maybe not as applicable to this area folks?

tehhowch, tibetiroka, vitalchip, EjoThims, TheMarksman, Anarchist2, Pointedstick, W1zrad

I know I don't strictly need any kind of approval to put up a PR, but I'd at least like to get some level of "real insight into the specifics" from those that will ultimately decide if something can get merged or not. At a minimum to make sure what I'm proposing aligns with the goals of the project and that there isn't some major change they would like to see now before I do any significant implementation.

Also a small note about the labels, this got labeled with "mechanics" but I would specifically be endeavoring not to change any game mechanics with anything I've proposed here. Maybe that was automatic by virtue of the category, but I think "enhancement" and "refactor" would be the most appropriate labels.

**Reply by Zitchas** [reply]
I'm mostly on the content/lore/story side of things, my limited code experience is improving the UI and information available to the player. 

Tehhowch is really the one who has the best knowledge of the code, but they've had Real Life stuff take up a lot of their time lately. I'm not sure which of our coders have the most knowledge/interest in input systems. Hurleveur, Samrocketman, Amazinite, and Warp-core are probably the main coders that are active and around currently.

**Comment by Saugia** [comment]
Amazinite is the lead dev of Endless Sky. Warp-Core, Peter, and Quyykk are our most active Code Dev's. You're getting the eyes of those you seek, but it might be a hot moment before responses.

**Comment by oo13** [comment]
SDL has [SDL_TextInputEvent](https://wiki.libsdl.org/SDL2/SDL_TextInputEvent), which supports some exotic keyboards, including French.
(Cf. [the comment](https://github.com/endless-sky/endless-sky/issues/558#issuecomment-622351133) and [the commit](https://github.com/Hagb/endless-sky/commit/78877147458f5e49c43246609649106be7b6826b). It might be complex because it supports IME.)

I'm glad the information helps you.

**Reply by Zireael07** [reply]
Why is this suddenly marked as closed? I haven't seen any solutions proposed nor merged

**Reply by warp-core** [reply]
GitHub says the author, @gnab-gib-se, did it.
